stan/math/opencl/kernel_generator.hpp

```diff
@@ -2,6 +2,107 @@
 #define STAN_MATH_OPENCL_KERNEL_GENERATOR_HPP
 #ifdef STAN_OPENCL
 
+/**
+ * \ingroup opencl
+ * \defgroup opencl_kernel_generator OpenCL Kernel Generator
+ *
+ * The OpenCL kernel generator is used to combine multiple matrix operations
+ * into a single OpenCL kernel. This is much simpler than writing
+ * multi-operation kernels by hand.
+ *
+ * Because global GPU memory loads and stores are relativly slow compared to
+ * calculations in a kernel, using one kernel for multiple operations is faster
+ * than using one kernel per operation.
+ *
+ * The kernel generator uses lazy evaluation. Each operation is represented by
+ * an object derived from `operation_cl`. Such an object holds arguments of the
+ * operations as well as meta information needed to generate calculations on the
+ * arguments. Arguments to operations can be other operations, scalars
+ * or `matrix_cl` objects. An operation is evaluated when either an operation is
+ * assigned to a `matrix_cl` or a left-hand-side operation or `.eval()` is
+ * called.
+ *
+ * ## Defining a new kernel generator operation
+ *
+ * New kernel generator classes must satsify the conditions below:
```

Comment:
```suggestion
 * ## Defining a new kernel generator expression
 *
 * For adding new Unary and Binary ops use the macros provided in
 * `unary_function_cl.hpp` and `binary_operation.hpp`.
 *
 * New kernel generator expressions classes must satisfy the conditions below:
```

Reply:
Macro for binary operation only works for OpenCL operators. I think I covered all of them.